fed/cleanup.py

```diff
@@ -32,75 +33,145 @@ class CleanupManager:
+                        downstream_seq_id: int = -1,
+                        is_error: bool = False):
+        """
+        Push the sending remote task's return value, i.e. `obj_ref` to
+        the corresponding message queue.
+
+        Args:
+            obj_ref: The return value of the send remote task.
+            dest_party: Destination
+            upstream_seq_id: (Optional) This is unneccessary when sending
+                normal data message because it was already sent to the target
+                party. However, if the computation is corrupted, an error object
+                will be created and sent with the same seq_id to replace the
+                original data object. This argument is used to send the error.
+            downstream_seq_id: (Optional) Same as `upstream_seq_id`.
+            is_error: (Optional) Whether the obj_ref represent an error object or not.
+                Default to False. If True, the obj_ref will be sent to the error
+                queue instead.
+        """
+        msg_pack = (obj_ref, dest_party, upstream_seq_id, downstream_seq_id)
+        if (is_error):
+            self._sending_error_q.push(msg_pack)
+        else:
+            self._sending_data_q.push(msg_pack)
+
+    def _signal_exit(self):
+        """
+        Exit the current process immediately. The signal will be captured
+        in main thread where the `stop` will be called.
+        """
+        # NOTE(NKcqx): The signal is implemented by the error mechanism,
+        # a `KeyboardInterrupt` will be raised after sending the signal,
+        # and OS will hold
+        # the process's original context and change to the error handler context.
+        # States that the original context hold, including `threading.lock`,
+        # will not be released, acquiring the same lock in signal handler
+        # will cause dead lock. In order to ensure executing `shutdown` exactly
+        # once and avoid dead lock, the lock must be checked before sending
+        # signals.
+        if (self._acquire_shutdown_flag()):
+            logger.debug("Signal SIGINT to exit.")
+            os.kill(os.getpid(), signal.SIGINT)
+
+    def _process_data_message(self, message):
+        """
+        This is the handler function for message queue and will be called
+        in the sub-thread.
+        """
+        obj_ref, dest_party, upstream_seq_id, downstream_seq_id = message
+        try:
+            res = ray.get(obj_ref)
+        except Exception as e:
+            logger.warn(f'Failed to send {obj_ref} to {dest_party}, error: {e},'
+                        f'upstream_seq_id: {upstream_seq_id}, '
+                        f'downstream_seq_id: {downstream_seq_id}.')
+            if (isinstance(e, RayError)):
+                logger.info(f"Sending error {e.cause} to {dest_party}.")
+                from fed.proxy.barriers import send
+                # TODO(NKcqx): Maybe broadcast to all parties?
```

Comment:
I believe we have to broadcast to all parties, for example A -> B -> C，A triggers error then A and B exit, however C will hang forever?

Reply:
I'll start another PR to implement, since this PR is already hard to understand